providers/stores/abstractFileStore.js

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation and others. Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+const utils = require('../../lib/utils')
+const fs = require('fs')
+const path = require('path')
+const recursive = require('recursive-readdir')
+const { promisify } = require('util')
+const ResultCoordinates = require('../../lib/resultCoordinates')
+
+class AbstractFileStore {
+  constructor(options) {
+    this.options = options
+  }
+
+  async initialize() {}
+
+  /**
+   * Visit all of the files associated with the given coordinates.
+   *
+   * @param {EntityCoordinates} coordinates - Accepts partial coordinates.
+   * @returns The collection of results returned by the visitor
+   */
+  async list(coordinates, visitor) {
+    try {
+      const paths = await recursive(this._toStoragePathFromCoordinates(coordinates), ['.DS_Store'])
+      return await Promise.all(
+        paths.map(async path => {
+          if (!this._isValidPath(path)) return null
+          const data = await promisify(fs.readFile)(path)
+          return data ? visitor(JSON.parse(data)) : null
+        })
+      )
+    } catch (error) {
+      // If there is just no entry, that's fine, there is no content.
+      if (error.code === 'ENOENT') return []
+      throw error
+    }
+  }
+
+  /**
+   * Get and return the object at the given coordinates.
+   *
+   * @param {Coordinates} coordinates - The coordinates of the object to get
+   * @returns The loaded object
+   */
+  async get(coordinates) {
+    const filePath = await this._getFilePath(coordinates)
+    try {
+      const result = await promisify(fs.readFile)(filePath)
+      return JSON.parse(result)
+    } catch (error) {
+      if (error.code === 'ENOENT') return null
+      throw error
+    }
+  }
+
+  _isValidPath(entry) {
+    return AbstractFileStore.isInterestingCoordinates(this._toResultCoordinatesFromStoragePath(entry))
+  }
+
+  _toStoragePathFromCoordinates(coordinates) {
+    const result = AbstractFileStore.toStoragePathFromCoordinates(coordinates)
+    return path.join(this.options.location, result).replace(/\\/g, '/')
+  }
+
+  _toResultCoordinatesFromStoragePath(path) {
+    const trimmed = path.slice(this.options.location.length + 1)
+    return AbstractFileStore.toResultCoordinatesFromStoragePath(trimmed)
+  }
+
+  async _getFilePath(coordinates) {
+    const toolPath = this._toStoragePathFromCoordinates(coordinates)
+    if (coordinates.toolVersion) return toolPath + '.json'
+    const latest = await this._findLatest(toolPath)
+    if (!latest) return null
+    return path.join(toolPath, latest) + '.json'
```

Comment:
do you think this belongs in the AbstractFileStore? It's very tied to tool results.

It may make more sense in `fileHarvestStore.js`